### PR TITLE
#1162: applied refactorings

### PR DESCRIPTION
**Pull request by jan-vcapgemini**
Addresses: #1162 #1166 

Imlements:
* refactored checkDownloadUrl #1166 
* added isContentValid and isChecksumStillValid methods added new doAddVersion without checksum
* removed emtpy string from old doAddVersion calls using os and architecture
* removed checksum check from doAddVersion
* removed 2nd checkDownloadUrl method
* renamed UpdateManagerIT to JsonUrlUpdaterIT
* replaced base URLs with localhost
* added test and resource for missing checksums
* added test for missing downloads
* renamed tests

### CODE REVIEW

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:207]: this change is incorrect and the old Javadoc of `urlVersion` was correct and more appropriate. Was this on accident or on purpose?

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:224]: IMHO this code does not make sense:
The getter will return what has been given to the setter so `checksum` and `oldChecksum` will always refer to the same value.
IMHO the entire method `isChecksumStillValid` does not make sense yet. If we would have a method with that name then either to verify the checksum for the end-user but that code is currently still in bash or in case we would revalidate this on update of existing versions. However, due to the greediness of checksum computation this is not planned.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:244]: maybe `isValidDownload` would be more suitable as `text/html` in general is a valid content-type but not for a binary download. Actually we would expect `application/octet-stream` but not every server is 100% compliant so I better went for a blacklist than a whitelist.

**Review by hohwille — CHANGES_REQUESTED** [review]
@jan-vcapgemini Thanks for your PR. Great that you cleaned the signature change and awesome that you created excellent integration tests. 👍 
Can you have a look at my review comments and update accordingly?

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:244]: You could even pass the response as parameter and read the content type within this method, what would make more even sense. Thanks for extracting such smaller methods to make the code better!

Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:244]: Maybe even one step further, you could then also do the `isSuccess` check in this method and simplify `checkDownloadUrl` further.

File: url-updater/src/test/java/com/devonfw/tools/ide/integrationtests/JsonUrlUpdaterIT.java
Comment (hohwille) [inline url-updater/src/test/java/com/devonfw/tools/ide/integrationtests/JsonUrlUpdaterIT.java:125]: is this the SHA256 sum of "aBody"? If so adding a comment or extracting to a named constant could help the reader of the code.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (jan-vcapgemini) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: I've implemented all of your mentioned points. Thanks for your constructive feedback. I hope that it looks better now.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (jan-vcapgemini) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: I've adjusted it now.

File: url-updater/src/test/java/com/devonfw/tools/ide/integrationtests/JsonUrlUpdaterIT.java
Comment (jan-vcapgemini) [reply url-updater/src/test/java/com/devonfw/tools/ide/integrationtests/JsonUrlUpdaterIT.java]: I've extracted a named constant now.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (jan-vcapgemini) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: I've removed the addUrl and setChecksum now. Not sure if we should remove the whole method though.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:279]: This line is still redundant and could be moved to a private method for reuse.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:270]: It would be even smarter to distinguish if we already have a checksum or not:
* if we already have a checksum given, a HEAD request is sufficient
* otherwise we need the full download and can do only a GET - currently you always do HEAD and then additionally to a GET request what is overhead.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [inline url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:344]: thanks for also fixing typos :)

**Review by hohwille — APPROVED** [review]
@jan-vcapgemini thanks for your rework. This PR is now great and so far correct. 👍 
I again left some nice-to-have comments for the future but I want to get this merged to resolve the conflicts due to the signature changes.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:270]: We will do this after the merge to not block the merge.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java:279]: this is a really picky nice-to-have that can be done after the merge.

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: Thanks. See my new comment for further improvements...

File: url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java
Comment (hohwille) [reply url-updater/src/main/java/com/devonfw/tools/ide/url/updater/AbstractUrlUpdater.java]: We can discuss this after merge. However, the method still seems pointless to me: if it is called from `addVersion` the version is new and no previous checksum can exist. If we use this method on update of versions this makes sense but in such case we never set the checksum, as we already assume that it cannot change afterwards - a release by design has to be immutable.

**Pull request merged**
